core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1770,45 +1770,47 @@ class KafkaApis(val requestChannel: RequestChannel,
       sendResponseMaybeThrottle(controllerMutationQuota, request, createResponse, onComplete = None)
     }
 
+    // be sure to check authorization first, before checking if this is the controller, to avoid leaking
+    // information about the system (i.e. who is the controller) to principals unauthorized for that information
+
     val createTopicsRequest = request.body[CreateTopicsRequest]
     val results = new CreatableTopicResultCollection(createTopicsRequest.data.topics.size)
-    if (!controller.isActive) {
-      createTopicsRequest.data.topics.forEach { topic =>
-        results.add(new CreatableTopicResult().setName(topic.name)
-          .setErrorCode(Errors.NOT_CONTROLLER.code))
-      }
-      sendResponseCallback(results)
-    } else {
-      createTopicsRequest.data.topics.forEach { topic =>
-        results.add(new CreatableTopicResult().setName(topic.name))
+    createTopicsRequest.data.topics.forEach { topic =>
+      results.add(new CreatableTopicResult().setName(topic.name))
+    }
+    val hasClusterAuthorization = authorize(request.context, CREATE, CLUSTER, CLUSTER_NAME,
+      logIfDenied = false)
+    val topics = createTopicsRequest.data.topics.asScala.map(_.name)
+    val authorizedTopics =
+      if (hasClusterAuthorization) topics.toSet
+      else filterByAuthorized(request.context, CREATE, TOPIC, topics)(identity)
+    val authorizedForDescribeConfigs = filterByAuthorized(request.context, DESCRIBE_CONFIGS, TOPIC,
+      topics, logIfDenied = false)(identity).map(name => name -> results.find(name)).toMap
+
+    results.forEach { topic =>
+      if (results.findAll(topic.name).size > 1) {
+        topic.setErrorCode(Errors.INVALID_REQUEST.code)
+        topic.setErrorMessage("Found multiple entries for this topic.")
+      } else if (!authorizedTopics.contains(topic.name)) {
+        topic.setErrorCode(Errors.TOPIC_AUTHORIZATION_FAILED.code)
+        topic.setErrorMessage("Authorization failed.")
+      }
+      if (!authorizedForDescribeConfigs.contains(topic.name)) {
+        topic.setTopicConfigErrorCode(Errors.TOPIC_AUTHORIZATION_FAILED.code)
       }
-      val hasClusterAuthorization = authorize(request.context, CREATE, CLUSTER, CLUSTER_NAME,
-        logIfDenied = false)
-      val topics = createTopicsRequest.data.topics.asScala.map(_.name)
-      val authorizedTopics =
-        if (hasClusterAuthorization) topics.toSet
-        else filterByAuthorized(request.context, CREATE, TOPIC, topics)(identity)
-      val authorizedForDescribeConfigs = filterByAuthorized(request.context, DESCRIBE_CONFIGS, TOPIC,
-        topics, logIfDenied = false)(identity).map(name => name -> results.find(name)).toMap
-
+    }
+    if (!controller.isActive) {
```

Comment:
I think we could probably roll this into the loop above. What I'm thinking is something like this:
```scala
    val toCreate = mutable.Map[String, CreatableTopic]()

    results.forEach { topic =>
      if (results.findAll(topic.name).size > 1) {
        topic.setErrorCode(Errors.INVALID_REQUEST.code)
        topic.setErrorMessage("Found multiple entries for this topic.")
      } else if (!authorizedTopics.contains(topic.name)) {
        topic.setErrorCode(Errors.TOPIC_AUTHORIZATION_FAILED.code)
        topic.setErrorMessage("Authorization failed.")
      } else if (!controller.isActive) {
        topic.setErrorCode(Errors.NOT_CONTROLLER.code)
      } else {
        toCreate += topic.name -> topic
      }
      if (!authorizedForDescribeConfigs.contains(topic.name)) {
        topic.setTopicConfigErrorCode(Errors.TOPIC_AUTHORIZATION_FAILED.code)
      }
   }

   if (toCreate.isEmpty) {
     sendResponseCallback(results)
   } else {
     ...
```